### PR TITLE
AsRef/AsMut for `ipld_create/read`

### PR DESCRIPTION
**Pull request by mriise**
A little nicety I thought of while building kernel tests, not sure how it would relate to morphization bloat mentioned in #373

### CODE REVIEW

**Comment by Stebalien** [comment]
So, there's a trade-off here:

- Complexity of implementation.
- Complexity of use.

In general:

- For public interfaces, common concrete types, collections, etc .where I'm going to implement them once and use them frequently, I go full generic so that the caller can just pass in whatever they want.
- For common interfaces, or interfaces that only have a few users (not including tests), I optimize to make them easy to understand/read/implement, even if I need a little extra boilerplate to _use_ them.

In this case, given that the extra boilerplate to use them is a simple `&`, I'd leave the code as-is.

The _other_ consideration here is object safety, but I don't think that's _that_ much of an issue right now.


> would relate to morphization bloat mentioned in https://github.com/filecoin-project/ref-fvm/issues/373

That issue is specifically concerned with monomorphizeation in actors. The kernel lives in the FVM, so it's not an issue.

File: fvm/src/kernel/default.rs [hunk 1]
```diff
@@ -275,11 +275,11 @@ where
         Ok((id, stat))
     }
 
-    fn block_create(&mut self, codec: u64, data: &[u8]) -> Result<BlockId> {
+    fn block_create(&mut self, codec: u64, data: &(impl AsRef<[u8]> + ?Sized)) -> Result<BlockId> {
         self.call_manager
```
Comment (Stebalien) [inline fvm/src/kernel/default.rs:279]: note: one would usually call `let data = data.as_ref()` at the top.

File: fvm/src/kernel/default.rs [hunk 2]
```diff
@@ -275,11 +275,11 @@ where
...
-            .charge_gas(self.call_manager.price_list().on_block_create(data.len()))?;
+            .charge_gas(self.call_manager.price_list().on_block_create( core::mem::size_of_val(data.as_ref())))?;
```
Comment (Stebalien) [inline fvm/src/kernel/default.rs:280]: `len` is still the correct thing to call here. _Technically_, `size_of_val` will work, but it has different semantics (byte-size of an arbitrary dynamically sized type, versus the length of a byte buffer).

> core

Given that we depend on `std`, just use `std`. Use `core` in `no_std` crates.

File: fvm/src/kernel/default.rs [hunk 3]
```diff
@@ -314,19 +314,19 @@ where
         Ok(k)
     }
 
-    fn block_read(&mut self, id: BlockId, offset: u32, buf: &mut [u8]) -> Result<i32> {
+    fn block_read(&mut self, id: BlockId, offset: u32, buf: &mut (impl AsMut<[u8]> + ?Sized)) -> Result<i32> {
```
Comment (Stebalien) [inline fvm/src/kernel/default.rs:317]: One would usually take `impl AsRef<[u8]>`:

- This makes it possible to pass a buffer by-value (if desired).
- `AsRef<U>` is implemented over `&T` where `T: AsRef<U> +?Sized`.
- This lets you drop the `?Sized`.

**Review by Stebalien — COMMENTED** [review]
(comments just to teach a bit of rust)

File: fvm/src/kernel/default.rs (hunk 1 above)
Comment (Stebalien) [reply fvm/src/kernel/default.rs:279]: Why:

- Can leave the generics out of the rest of the function.
- Maybe the compiler will get smart and split the function into a monomorphized portion, and a shared portion.

File: fvm/src/kernel/default.rs (hunk 3 above)
Comment (Stebalien) [reply fvm/src/kernel/default.rs:317]: I.e.

```rust
     fn block_read(&mut self, id: BlockId, offset: u32, buf: impl AsMut<[u8]>) -> Result<i32> { 
```

**Comment by mriise** [comment]
> The other consideration here is object safety

good to know :+1:

**Comment by mriise** [comment]
closing since this won't be merged

**Pull request closed without merging**